### PR TITLE
Fix some memory leaks on panicking destructors

### LINKED ISSUE #26: Memory leaks when destructors panic
For more info see the branch I’ve already mentioned in #24, i.e.
[steffahn/self_cell@`fix_drop_order...fix_double_free_and_leaks`](https://github.com/Voultapher/self_cell/compare/ebbf355...c482860#diff-1778ecc1b76a790dccfe3d31d658c07d2464493912a0e873cce3cdf1a361f8d7R810-R854)
in particular, the highlighted test cases.

(FYI, that branch currently also contains #25)

These memory leaks are not a soundness issue; still, fixing the problem is a nice-to-have improvement.

I can rebase, resolve merge conflicts, and open a PR once #21 and #25 are merged.

### PR DESCRIPTION
**Pull request by steffahn**
Closes #26

### CODE REVIEW

**Comment by Voultapher** [comment]
Having both owner and dependent impl panic, not sure how to test that without running into terminate.

**Comment by steffahn** [comment]
If there's two panics then an abort is expected.

This could be tested by having the owner and the dependent implement drop, the dependent panicking, the owner's panic with some side-effect. Then catch the panic, and test for the side-effect.

I don't *really* think such a test is strictly necessary, in particular I don't think it would need to be part of this PR. 

Then you'll know that both destructors (owner and dependent) were run. Since owner was run during unwinding, if it were to panic an abort would've automatically resulted.

File: src/unsafe_self_cell.rs [hunk 1]
```diff
@@ -149,16 +156,31 @@ impl<Owner, Dependent> OwnerAndCellDropGuard<Owner, Dependent> {
 
 impl<Owner, Dependent> Drop for OwnerAndCellDropGuard<Owner, Dependent> {
     fn drop(&mut self) {
+        struct DeallocGuard {
```
Comment (Voultapher) [inline src/unsafe_self_cell.rs:159]: I heard you like drop guards so I put a drop guard inside your drop guard :D

**Review by Voultapher — APPROVED** [review]
LGTM

**Comment by steffahn** [comment]
Actually, I'm only now reading your modified tests. By making the owner a string, pure already testing whether the owner is dropped if dependent panics. If it wasn't, there would be a leak. Since it is dropped, we can be sure that if it's destructor panicked, an abort would be triggered.

**Comment by Voultapher** [comment]
@steffahn are we good for v0.9.3 or do you want to do some more research? IMO a day or two are not critical, but notifying users of an important update, I'd prefer doing once instead of twice or more.

**Comment by steffahn** [comment]
I'm not planning on doing anything else today, and as of now I'm not aware of any remaining unsoundness. I don't know yet if I'll take another look around the source tomorrow or not.

**Comment by Voultapher** [comment]
Ok, thanks for taking a look. I want this project to be easy and safe-to-use, without caveats. So I'm always happy to see soundness issues found and addressed. Better now than later. I'm quite busy right now myself. So I'll release a new version hopefully Friday and open a PR for bracket.

**Pull request merged**